lightly/loss/swav_loss.py

```diff
@@ -22,23 +27,33 @@ def sinkhorn(out: torch.Tensor, iterations: int = 3, epsilon: float = 0.05):
             Number of sinkhorn iterations.
         epsilon:
             Temperature parameter.
+        gather_distributed:
+            If True then features from all gpus are gathered to calculate the
+            soft codes Q. 
 
     Returns:
         Soft codes Q assigning each feature to a prototype.
     
     """
+    world_size = 1
+    if gather_distributed and dist.is_initialized():
+        world_size = dist.get_world_size()
 
     # get the exponential matrix and make it sum to 1
     Q = torch.exp(out / epsilon).t()
     sum_Q = torch.sum(Q)
+    if world_size > 1:
+        dist.all_reduce(sum_Q)
     Q /= sum_Q
 
-    B = Q.shape[1]
+    B = Q.shape[1] * world_size
     K = Q.shape[0] # number of prototypes
 
-    for i in range(iterations):
+    for _ in range(iterations):
         # normalize rows
         sum_of_rows = torch.sum(Q, dim=1, keepdim=True)
+        if world_size > 1:
+            dist.all_reduce(sum_of_rows)
         Q /= sum_of_rows
         Q /= K
```

Comment:
This line might be redundant.

Reply:
Q /= K is redundant, as the columns are normalised in the line below.

Reply:
If you insert `Q = Q / K` into `Q = Q / torch.sum(Q, dim=0, keepdim=True)` then you have
`Q = (Q / K) / torch.sum((Q / K), dim=0, keepdim=True)` and the K cancels itself out.